File: touristapp_backend/app/services/recommendation_service.py
from motor.motor_asyncio import AsyncIOMotorDatabase
from app.database import get_db
from app.models.user import UserInDB
from bson import ObjectId
from datetime import datetime, timedelta
from typing import List, Dict, Any, Optional
import random
import math
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.cluster import KMeans
import pandas as pd
from app.services.maps_service import calculate_distance
import asyncio

class RecommendationEngine:
    """Advanced recommendation engine for tourist spots and tour packages"""
    
    @staticmethod
    async def get_personalized_spots(
        user: UserInDB, 
        limit: int = 10,
        user_location: Optional[Dict[str, float]] = None
    ) -> List[Dict[str, Any]]:
        """Get personalized tourist spot recommendations for a user"""
        db = get_db()
        spots_collection = db["tourist_spots"]
        bookings_collection = db["bookings"]
        ratings_collection = db["ratings"]
        
        # Get user's booking and rating history
        user_history = await RecommendationEngine._get_user_history(user.id)
        
        # Get all tourist spots
        all_spots = await spots_collection.find({}).to_list(length=None)
        
        recommendations = []
        for spot_data in all_spots:
            score = await RecommendationEngine._calculate_spot_score(
                spot_data, user, user_history, user_location
            )
            
            if score > 0:
                spot_dict = {
                    "id": str(spot_data["_id"]),
                    "name": spot_data["name"],
                    "description": spot_data["description"],
                    "location": spot_data["location"],
                    "region": spot_data["region"],
                    "categories": spot_data["categories"],
                    "rating": spot_data.get("rating", 0.0),
                    "total_ratings": spot_data.get("total_ratings", 0),
                    "entry_fee": spot_data.get("entry_fee", 0.0),
                    "image_urls": spot_data.get("image_urls", []),
                    "recommendation_score": round(score, 2),
                    "recommendation_reasons": RecommendationEngine._get_recommendation_reasons(
                        spot_data, user, user_history
                    )
                }
                
                # Add distance if user location provided
                if user_location and "location" in spot_data:
                    distance = calculate_distance(
                        user_location["latitude"], user_location["longitude"],
                        spot_data["location"]["latitude"], spot_data["location"]["longitude"]
                    )
                    spot_dict["distance_km"] = round(distance, 2)
                
                recommendations.append(spot_dict)
        
        # Sort by recommendation score and return top results
        recommendations.sort(key=lambda x: x["recommendation_score"], reverse=True)
        return recommendations[:limit]
    
    @staticmethod
    async def get_personalized_packages(
        user: UserInDB, 
        limit: int = 10
    ) -> List[Dict[str, Any]]:
        """Get personalized tour package recommendations"""
        db = get_db()
        packages_collection = db["tour_packages"]
        
        user_history = await RecommendationEngine._get_user_history(user.id)
        all_packages = await packages_collection.find({"status": "active"}).to_list(length=None)
        
        recommendations = []
        for package_data in all_packages:
            score = await RecommendationEngine._calculate_package_score(
                package_data, user, user_history
            )
            
            if score > 0:
                package_dict = {
                    "id": str(package_data["_id"]),
                    "name": package_data["name"],
                    "description": package_data["description"],
                    "price": package_data["price"],
                    "duration_days": package_data["duration_days"],
                    "destinations": package_data["destinations"],
                    "max_participants": package_data["max_participants"],
                    "current_participants": package_data.get("current_participants", 0),
                    "recommendation_score": round(score, 2),
                    "recommendation_reasons": RecommendationEngine._get_package_recommendation_reasons(
                        package_data, user, user_history
                    )
                }
                recommendations.append(package_dict)
        
        recommendations.sort(key=lambda x: x["recommendation_score"], reverse=True)
        return recommendations[:limit]
    
    @staticmethod
    async def get_similar_spots(spot_id: str, limit: int = 5) -> List[Dict[str, Any]]:
        """Get spots similar to a given spot"""
        db = get_db()
        spots_collection = db["tourist_spots"]
        
        # Get the reference spot
        reference_spot = await spots_collection.find_one({"_id": ObjectId(spot_id)})
        if not reference_spot:
            return []
        
        # Find similar spots based on categories and region
        similar_spots = []
        all_spots = await spots_collection.find({
            "_id": {"$ne": ObjectId(spot_id)}
        }).to_list(length=None)
        
        for spot_data in all_spots:
            similarity_score = RecommendationEngine._calculate_spot_similarity(
                reference_spot, spot_data
            )
            
            if similarity_score > 0.3:  # Threshold for similarity
                spot_dict = {
                    "id": str(spot_data["_id"]),
                    "name": spot_data["name"],
                    "description": spot_data["description"],
                    "location": spot_data["location"],
                    "categories": spot_data["categories"],
                    "rating": spot_data.get("rating", 0.0),
                    "similarity_score": round(similarity_score, 2)
                }
                similar_spots.append(spot_dict)
        
        similar_spots.sort(key=lambda x: x["similarity_score"], reverse=True)
        return similar_spots[:limit]
    
    @staticmethod
    async def get_trending_spots(limit: int = 10) -> List[Dict[str, Any]]:
        """Get trending tourist spots based on recent bookings and ratings"""
        db = get_db()
        spots_collection = db["tourist_spots"]
        bookings_collection = db["bookings"]
        ratings_collection = db["ratings"]
        
        # Get spots with high ratings and recent activity
        pipeline = [
            {
                "$match": {
                    "rating": {"$gte": 4.0},
                    "total_ratings": {"$gte": 5}
                }
            },
            {
                "$sort": {
                    "rating": -1,
                    "total_ratings": -1
                }
            },
            {
                "$limit": limit
            }
        ]
        
        trending_spots = await spots_collection.aggregate(pipeline).to_list(length=None)
        
        result = []
        for spot_data in trending_spots:
            spot_dict = {
                "id": str(spot_data["_id"]),
                "name": spot_data["name"],
                "description": spot_data["description"],
                "location": spot_data["location"],
                "categories": spot_data["categories"],
                "rating": spot_data.get("rating", 0.0),
                "total_ratings": spot_data.get("total_ratings", 0),
                "trending_score": spot_data.get("rating", 0.0) * (1 + spot_data.get("total_ratings", 0) / 100)
            }
            result.append(spot_dict)
        
        return result
    
    @staticmethod
    async def _get_user_history(user_id: str) -> Dict[str, Any]:
        """Get user's booking and rating history"""
        db = get_db()
        bookings_collection = db["bookings"]
        ratings_collection = db["ratings"]
        
        # Get user bookings
        bookings = await bookings_collection.find({"user_id": user_id}).to_list(length=None)
        
        # Get user ratings
        ratings = await ratings_collection.find({"tourist_id": user_id}).to_list(length=None)
        
        return {
            "bookings": bookings,
            "ratings": ratings,
            "visited_categories": RecommendationEngine._extract_categories_from_history(bookings),
            "preferred_price_range": RecommendationEngine._extract_price_preferences(bookings),
            "average_rating_given": RecommendationEngine._calculate_average_rating(ratings)
        }
    
    @staticmethod
    async def _calculate_spot_score(
        spot_data: Dict[str, Any], 
        user: UserInDB, 
        user_history: Dict[str, Any],
        user_location: Optional[Dict[str, float]] = None
    ) -> float:
        """Calculate recommendation score for a tourist spot"""
        score = 0.0
        
        # Base score from spot rating
        score += spot_data.get("rating", 0.0) * 0.3
        
        # Category preference matching
        if user.preferences and user.preferences.preferred_categories:
            spot_categories = [cat.lower() for cat in spot_data.get("categories", [])]
            user_categories = [cat.lower() for cat in user.preferences.preferred_categories]
            
            category_match = len(set(spot_categories) & set(user_categories))
            score += category_match * 2.0
        
        # Region preference matching
        if user.preferences and user.preferences.preferred_regions:
            if spot_data.get("region", "").lower() in [r.lower() for r in user.preferences.preferred_regions]:
                score += 1.5
        
        # Historical category preferences
        if user_history["visited_categories"]:
            spot_categories = set(cat.lower() for cat in spot_data.get("categories", []))
            history_categories = set(user_history["visited_categories"])
            
            if spot_categories & history_categories:
                score += 1.0
        
        # Distance factor (closer is better if location provided)
        if user_location and "location" in spot_data:
            distance = calculate_distance(
                user_location["latitude"], user_location["longitude"],
                spot_data["location"]["latitude"], spot_data["location"]["longitude"]
            )
            # Reduce score for very distant spots
            if distance > 100:
                score *= 0.5
            elif distance > 50:
                score *= 0.8
        
        # Budget consideration
        entry_fee = spot_data.get("entry_fee", 0.0)
        if user.preferences and user.preferences.budget_range:
            budget_max = user.preferences.budget_range.get("max", 1000)
            if entry_fee <= budget_max:
                score += 0.5
            else:
                score *= 0.3  # Penalize expensive spots if over budget
        
        return max(0.0, score)
    
    @staticmethod
    async def _calculate_package_score(
        package_data: Dict[str, Any], 
        user: UserInDB, 
        user_history: Dict[str, Any]
    ) -> float:
        """Calculate recommendation score for a tour package"""
        score = 0.0
        
        # Budget matching
        package_price = package_data.get("price", 0)
        if user.preferences and user.preferences.budget_range:
            budget_min = user.preferences.budget_range.get("min", 0)
            budget_max = user.preferences.budget_range.get("max", 10000)
            
            if budget_min <= package_price <= budget_max:
                score += 3.0
            elif package_price > budget_max:
                score *= 0.2  # Heavy penalty for over-budget packages
        
        # Duration preferences (based on travel style)
        duration = package_data.get("duration_days", 1)
        if user.preferences and user.preferences.travel_style:
            if user.preferences.travel_style == "quick" and duration <= 3:
                score += 1.0
            elif user.preferences.travel_style == "extended" and duration >= 7:
                score += 1.0
            elif user.preferences.travel_style == "standard" and 3 <= duration <= 7:
                score += 1.0
        
        # Group size matching
        if user.preferences and user.preferences.group_size_preference:
            max_participants = package_data.get("max_participants", 1)
            if max_participants >= user.preferences.group_size_preference:
                score += 0.5
        
        # Historical preferences
        if user_history["preferred_price_range"]:
            hist_min, hist_max = user_history["preferred_price_range"]
            if hist_min <= package_price <= hist_max:
                score += 1.0
        
        return max(0.0, score)
    
    @staticmethod
    def _calculate_spot_similarity(spot1: Dict[str, Any], spot2: Dict[str, Any]) -> float:
        """Calculate similarity between two spots"""
        similarity = 0.0
        
        # Category similarity
        cats1 = set(spot1.get("categories", []))
        cats2 = set(spot2.get("categories", []))
        
        if cats1 and cats2:
            category_similarity = len(cats1 & cats2) / len(cats1 | cats2)
            similarity += category_similarity * 0.6
        
        # Region similarity
        if spot1.get("region") == spot2.get("region"):
            similarity += 0.3
        
        # Rating similarity
        rating1 = spot1.get("rating", 0.0)
        rating2 = spot2.get("rating", 0.0)
        rating_diff = abs(rating1 - rating2)
        rating_similarity = max(0, 1 - rating_diff / 5.0)
        similarity += rating_similarity * 0.1
        
        return similarity
    
    @staticmethod
    def _extract_categories_from_history(bookings: List[Dict[str, Any]]) -> List[str]:
        """Extract preferred categories from booking history"""
        # This would require joining with tour packages to get destinations
        # For now, return empty list - would need more complex query
        return []
    
    @staticmethod
    def _extract_price_preferences(bookings: List[Dict[str, Any]]) -> Optional[tuple]:
        """Extract price preferences from booking history"""
        if not bookings:
            return None
        
        prices = [booking.get("total_amount", 0) for booking in bookings if booking.get("total_amount")]
        if not prices:
            return None
        
        return (min(prices), max(prices))
    
    @staticmethod
    def _calculate_average_rating(ratings: List[Dict[str, Any]]) -> float:
        """Calculate average rating given by user"""
        if not ratings:
            return 0.0
        
        total_rating = sum(rating.get("rating", 0) for rating in ratings)
        return total_rating / len(ratings)
    
    @staticmethod
    def _get_recommendation_reasons(
        spot_data: Dict[str, Any], 
        user: UserInDB, 
        user_history: Dict[str, Any]
    ) -> List[str]:
        """Generate human-readable reasons for recommendation"""
        reasons = []
        
        if spot_data.get("rating", 0) >= 4.5:
            reasons.append("Highly rated by other travelers")
        
        if user.preferences and user.preferences.preferred_categories:
            spot_categories = [cat.lower() for cat in spot_data.get("categories", [])]
            user_categories = [cat.lower() for cat in user.preferences.preferred_categories]
            
            matches = set(spot_categories) & set(user_categories)
            if matches:
                reasons.append(f"Matches your interest in {', '.join(matches)}")
        
        if user.preferences and user.preferences.preferred_regions:
            if spot_data.get("region", "").lower() in [r.lower() for r in user.preferences.preferred_regions]:
                reasons.append(f"Located in your preferred region: {spot_data.get('region')}")
        
        entry_fee = spot_data.get("entry_fee", 0.0)
        if entry_fee == 0:
            reasons.append("Free entry")
        elif entry_fee < 10:
            reasons.append("Affordable entry fee")
        
        if not reasons:
            reasons.append("Popular destination")
        
        return reasons
    
    @staticmethod
    def _get_package_recommendation_reasons(
        package_data: Dict[str, Any], 
        user: UserInDB, 
        user_history: Dict[str, Any]
    ) -> List[str]:
        """Generate reasons for package recommendations"""
        reasons = []
        
        if user.preferences and user.preferences.budget_range:
            package_price = package_data.get("price", 0)
            budget_max = user.preferences.budget_range.get("max", 10000)
            
            if package_price <= budget_max:
                reasons.append("Within your budget")
        
        duration = package_data.get("duration_days", 1)
        if duration <= 3:
            reasons.append("Perfect for a short getaway")
        elif duration >= 7:
            reasons.append("Great for an extended vacation")
        else:
            reasons.append("Ideal duration for a memorable trip")
        
        if package_data.get("current_participants", 0) < package_data.get("max_participants", 1):
            reasons.append("Available spots remaining")
        
        if not reasons:
            reasons.append("Popular tour package")
        
        return reasons

File: tourist_app/src/services/recommendationService.ts
import apiService from './apiService';
import { ApiResponse } from '../config/api';

// Ensure URLSearchParams is available
if (typeof URLSearchParams === 'undefined') {
  global.URLSearchParams = require('react-native-url-polyfill/auto').URLSearchParams;
}

export interface RecommendedSpot {
  id: string;
  name: string;
  description: string;
  location: {
    latitude: number;
    longitude: number;
    address: string;
  };
  category: string;
  rating: number;
  image_urls: string[];
  region: string;
  popularity_score?: number;
  distance?: number;
}

export interface RecommendedPackage {
  id: string;
  name: string;
  description: string;
  price: number;
  duration_days: number;
  rating: number;
  average_rating: number;
  total_ratings: number;
  image_urls: string[];
  category: string;
  difficulty_level: string;
  max_participants: number;
  current_participants: number;
  status: string;
}

class RecommendationService {
  // Get personalized tourist spot recommendations
  async getPersonalizedSpots(useAI: boolean = true, limit: number = 10): Promise<ApiResponse<RecommendedSpot[]>> {
    try {
      const response = await apiService.get<any>(`/recommendations/spots?use_ai=${useAI}&limit=${limit}`);

      if (response.success && response.data?.recommendations) {
        return {
          success: true,
          data: response.data.recommendations
        };
      }

      // If AI recommendations fail, fallback to popular spots
      if (!response.success && useAI) {
        console.log('AI recommendations failed, falling back to popular spots');
        return this.getPopularSpots(limit);
      }

      return response;
    } catch (error) {
      console.error('Error getting personalized spots:', error);
      // Fallback to popular spots
      return this.getPopularSpots(limit);
    }
  }

  // Get popular tourist spots as fallback
  async getPopularSpots(limit: number = 10): Promise<ApiResponse<RecommendedSpot[]>> {
    try {
      const queryParams = new URLSearchParams({
        limit: limit.toString(),
        sort: 'rating',
        order: 'desc'
      });

      const response = await apiService.get(`/tourist-spots/?${queryParams.toString()}`);

      if (response.success && response.data && Array.isArray(response.data)) {
        // Add popularity score based on rating and other factors
        const spotsWithScore = response.data.map((spot: any) => ({
          ...spot,
          popularity_score: this.calculatePopularityScore(spot)
        }));

        return {
          success: true,
          data: spotsWithScore.slice(0, limit)
        };
      }

      return {
        success: false,
        message: 'Failed to load popular spots'
      };
    } catch (error) {
      console.error('Error getting popular spots:', error);
      return {
        success: false,
        message: 'Failed to load popular spots'
      };
    }
  }

  // Get personalized package recommendations
  async getPersonalizedPackages(useAI: boolean = true, limit: number = 10): Promise<ApiResponse<RecommendedPackage[]>> {
    try {
      const response = await apiService.get<any>(`/recommendations/packages?use_ai=${useAI}&limit=${limit}`);

      if (response.success && response.data?.recommendations) {
        return {
          success: true,
          data: response.data.recommendations
        };
      }

      // If AI recommendations fail, fallback to featured packages
      if (!response.success && useAI) {
        console.log('AI package recommendations failed, falling back to featured packages');
        return this.getFeaturedPackages(limit);
      }

      return response;
    } catch (error) {
      console.error('Error getting personalized packages:', error);
      // Fallback to featured packages
      return this.getFeaturedPackages(limit);
    }
  }

  // Get featured packages as fallback
  async getFeaturedPackages(limit: number = 10): Promise<ApiResponse<RecommendedPackage[]>> {
    try {
      const queryParams = new URLSearchParams({
        limit: (limit * 2).toString(), // Get more to filter and sort
        status: 'active'
      });

      const response = await apiService.get(`/tour-packages/?${queryParams.toString()}`);

      if (response.success && response.data && Array.isArray(response.data)) {
        // Sort by rating and popularity to get featured packages
        const featuredPackages = response.data
          .filter((pkg: any) => pkg.status === 'active' || pkg.is_active)
          .sort((a: any, b: any) => {
            // Sort by average rating first, then by total ratings (popularity)
            const ratingA = a.average_rating || a.rating || 0;
            const ratingB = b.average_rating || b.rating || 0;

            if (ratingB !== ratingA) {
              return ratingB - ratingA;
            }

            const popularityA = (a.total_ratings || 0) + (a.current_participants || 0);
            const popularityB = (b.total_ratings || 0) + (b.current_participants || 0);

            return popularityB - popularityA;
          })
          .slice(0, limit);

        return {
          success: true,
          data: featuredPackages
        };
      }

      return {
        success: false,
        message: 'Failed to load featured packages'
      };
    } catch (error) {
      console.error('Error getting featured packages:', error);
      return {
        success: false,
        message: 'Failed to load featured packages'
      };
    }
  }

  // Get trending spots
  async getTrendingSpots(limit: number = 10): Promise<ApiResponse<RecommendedSpot[]>> {
    try {
      const response = await apiService.get<any>(`/recommendations/trending/spots?limit=${limit}`);

      if (response.success && response.data?.trending_spots) {
        return {
          success: true,
          data: response.data.trending_spots
        };
      }

      // If trending API fails, fallback to popular spots
      if (!response.success) {
        console.log('Trending spots API failed, falling back to popular spots');
        return this.getPopularSpots(limit);
      }

      return response;
    } catch (error) {
      console.error('Error getting trending spots:', error);
      return this.getPopularSpots(limit);
    }
  }

  // Get similar spots for a given spot
  async getSimilarSpots(spotId: string, limit: number = 5): Promise<ApiResponse<RecommendedSpot[]>> {
    try {
      const response = await apiService.get<any>(`/recommendations/spots/${spotId}/similar?limit=${limit}`);

      if (response.success && response.data?.similar_spots) {
        return {
          success: true,
          data: response.data.similar_spots
        };
      }

      // If similar spots API fails, get spots from same category
      if (!response.success) {
        console.log('Similar spots API failed, falling back to category-based recommendations');
        return this.getSpotsByCategory(spotId, limit);
      }

      return response;
    } catch (error) {
      console.error('Error getting similar spots:', error);
      return this.getSpotsByCategory(spotId, limit);
    }
  }

  // Fallback: Get spots from same category
  private async getSpotsByCategory(spotId: string, limit: number = 5): Promise<ApiResponse<RecommendedSpot[]>> {
    try {
      // First get the original spot to know its category
      const spotResponse = await apiService.get(`/tourist-spots/${spotId}`);

      if (!spotResponse.success || !spotResponse.data) {
        return {
          success: false,
          message: 'Could not find similar spots'
        };
      }

      const category = (spotResponse.data as any).category || '';

      // Get spots from same category
      const queryParams = new URLSearchParams({
        category,
        limit: (limit + 1).toString() // Get one extra to exclude the original
      });

      const categoryResponse = await apiService.get(`/tourist-spots/?${queryParams.toString()}`);

      if (categoryResponse.success && categoryResponse.data && Array.isArray(categoryResponse.data)) {
        // Filter out the original spot
        const similarSpots = categoryResponse.data
          .filter((spot: any) => spot.id !== spotId)
          .slice(0, limit);

        return {
          success: true,
          data: similarSpots
        };
      }

      return {
        success: false,
        message: 'Failed to load similar spots'
      };
    } catch (error) {
      console.error('Error getting spots by category:', error);
      return {
        success: false,
        message: 'Failed to load similar spots'
      };
    }
  }

  // Update user preferences for better recommendations
  async updateUserPreferences(preferences: {
    preferred_categories?: string[];
    budget_range?: { min: number; max: number };
    preferred_difficulty?: string;
    travel_style?: string;
  }): Promise<ApiResponse<any>> {
    try {
      return await apiService.post('/recommendations/update-preferences', preferences);
    } catch (error) {
      console.error('Error updating user preferences:', error);
      return {
        success: false,
        message: 'Failed to update preferences'
      };
    }
  }

  // Calculate popularity score for spots
  private calculatePopularityScore(spot: any): number {
    const rating = spot.rating || 0;
    const totalRatings = spot.total_ratings || 0;
    const recentViews = spot.recent_views || 0;

    // Weighted score: rating (40%) + total ratings (30%) + recent views (30%)
    return (rating * 0.4) + (Math.min(totalRatings / 10, 5) * 0.3) + (Math.min(recentViews / 100, 5) * 0.3);
  }

  // Search recommendations based on query
  async searchRecommendations(query: string, type: 'spots' | 'packages' = 'spots', limit: number = 10): Promise<ApiResponse<any[]>> {
    try {
      const queryParams = new URLSearchParams({
        search: query,
        limit: limit.toString()
      });

      if (type === 'spots') {
        return await apiService.get(`/tourist-spots/?${queryParams.toString()}`);
      } else {
        return await apiService.get(`/tour-packages/?${queryParams.toString()}`);
      }
    } catch (error) {
      console.error('Error searching recommendations:', error);
      return {
        success: false,
        message: 'Search failed'
      };
    }
  }
}

export const recommendationService = new RecommendationService();
export default recommendationService;